text/0009-BundleTracker-integration.md

```diff
@@ -425,33 +429,162 @@ Once a `BundleTracker` is opened, there are API methods to gain information and
+
+The customization scheme comes from a back pointer from the `TrackedBundle` class. When a `BundleTrackerCustomizer` is provided by the user, the pointer points to it. When `BundleTrackerCustomizer` is `nullptr` when constructed, it points to the `BundleTracker` itself to use the default or overridden methods.
+
+The constructor for `BundleTracker` takes a `std::shared_ptr` to a `BundleTrackerCustomizer` instance. This guarantees that the lifetime of the `BundleTrackerCustomizer` is scoped to that of the `BundleTracker` object. This is inconsistent with the current `ServiceTracker` specification, which uses a raw pointer. In the future, the `ServiceTracker` should be amended to follow this convention.
+
+It is worth noting that the API for `BundleTracker` is different from the current implementation of the `ServiceTracker`, which still uses a raw pointer for the `ServiceTracker` constructor's `ServiceTrackerCustomizer` argument. In the future, the `ServiceTracker` API should be adapted to function like the new `BundleTracker` API to make the lifetime of the `ServiceTrackerCustomizer` more explicit.
+### Core Framework Interactions
+
+The `BundleTracker` uses the following existing constructs in its implementation:
+
+- `BundleListener`: The listener is notified about `Bundle` state change events to be processed by the `BundleTracker`.
+- `BundleContext`: The context is used for registering the listener, and listing the known bundles to add during initialization.
+- `BundleAbstractTracked`: This class is used to manage the tracking map for the `ServiceTracker`. It is reused for the `BundleTracker`.
+
+The `BundleTracker` itself is part of the core CppMicroservices framework.
+
+### Class Design
+
+The following illustrates the class diagram for the `BundleTracker`:
+
+![Class Diagram](0009-BundleTracker-integration/class_diagram.svg)
+
+#### BundleTrackerCustomizer
+
+The `BundleTrackerCustomizer` is an interface that is defined in the API section. Since all of its methods are pure virtual, it contains no implementation and exists to be subclassed by `BundleTracker` and by users.
+
+#### BundleTracker
+
+The `BundleTracker` is what the user instantiates to track bundles. It contains a `std::unique_ptr` to a `BundleTrackerPrivate` object to store its data members.
+
+**Responsibilities**
+
+1. Provide tracking functionality to users
+2. Define default behavior of `AddingBundle()`, `ModifiedBundle()`, and `RemovedBundle()` callbacks.
+3. Create and manage the implementation constructs, like the `BundleTrackerPrivate`, `TrackedBundle`, and `BundleListener`.
 
+**Concurrency Note:**
+
+This class itself does not have locking since it contains no data, but grabs locks for `TrackedBundle` and `BundleTrackerPrivate` when doing synchronized work.
```

Comment:
Wording change:

**Before:** "This class itself does not have locking since ..."
**After:** "This class has no locks of its own since ..."

Reply:
This change has been adopted.